compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs

```diff
@@ -124,11 +124,22 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
                 // as that needs an exclusive borrow.
                 drop(coverage_map);
 
-                let coverageinfo = bx.tcx().coverageinfo(instance.def);
+                // The number of counters passed to `llvm.instrprof.increment` might
+                // be smaller than the number originally inserted by the instrumentor,
+                // if some high-numbered counters were removed by MIR optimizations.
+                let num_counters =
+                    bx.tcx().coverage_ids_info(instance.def).max_counter_id.as_u32() + 1;
+                if num_counters as usize > function_coverage_info.num_counters {
+                    // But it should never be larger.
+                    bug!(
+                        "num_counters disagreement: query says {num_counters} but function info only has {}",
+                        function_coverage_info.num_counters
+                    );
+                }
 
                 let fn_name = bx.get_pgo_func_name_var(instance);
                 let hash = bx.const_u64(function_coverage_info.function_source_hash);
-                let num_counters = bx.const_u32(coverageinfo.num_counters);
+                let num_counters = bx.const_u32(num_counters);
```

Comment:
I had originally wanted to do that, but eventually decided against it, for two reasons:

- LLVM uses this number to determine how many physical counter variables to create in memory and in `.profraw` files. So in cases where high-numbered counters are removed by MIR opts, this reduces the overall size of the profile data. Removing the query now would make profile data larger than before, in cases where it's currently making a difference.
- In the future, I want to improve this to get rid of *all* of the unused counters (instead of just the high-numbered ones), by renumbering the used counters to get rid of holes. If I remove the coverage query plumbing now, I'll just have to add it all back again later.